initial commit

lumber1000 opened the pull request.

lumber1000 committed:
initial commit

Nikita-Smirnov-Exactpro commented on build.gradle:
```diff
@@ -0,0 +1,98 @@
+
+plugins {
+    id "java"
+    id "org.jetbrains.kotlin.jvm" version "1.6.10"
```
Please use penultimate major release of Kotlin to avoid unknown problems of the last version

Nikita-Smirnov-Exactpro commented on build.gradle:
```diff
@@ -0,0 +1,98 @@
+    }
+}
+
+dependencies {
```
Please use slf4j and log4j implementation as part of your gentlemanly set for making a new project

Nikita-Smirnov-Exactpro commented on build.gradle:
```diff
@@ -0,0 +1,98 @@
+buildscript {
```
Include Gradle wrapper with a jar, executable files and properties in the project

Nikita-Smirnov-Exactpro commented on src/main/proto/simex_service.proto:
```diff
@@ -0,0 +1,72 @@
+  oneof event {
+    Trade trade = 11;
+    OrderAdded order_added_event = 12;
+    OrderSizeChanged order_size_changed = 13;
```
Ids of these fields are looked strange, use the reserved statement for removed fields or change ids before publication to a remote repository.
https://developers.google.com/protocol-buffers/docs/proto3#reserved

Nikita-Smirnov-Exactpro commented:
Implement CI via GitHub Actions 
https://docs.github.com/en/actions

Nikita-Smirnov-Exactpro commented on build.gradle:
```diff
@@ -0,0 +1,98 @@
+            grpckt {}
+        }
+    }
+    generatedFilesBaseDir = "$projectDir/src/main/generated"
```
Add the folder of generated files to project sourceSet

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Config.kt:
```diff
@@ -0,0 +1,14 @@
+import com.github.lumber1000.simex.service.SimexService
+
+const val HOSTNAME = "localhost"
+const val PORT = 50051
```
Pass connection params via cli or config file

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/BaseClient.kt:
```diff
@@ -0,0 +1,28 @@
+import io.grpc.ManagedChannelBuilder
+import com.github.lumber1000.simex.common.SimexServiceGrpc
+
+abstract class BaseClient(private val hostname: String, private val port: Int, protected val logger: Logger) {
```
Each class should create its own logger configured by its class. Information about the class can be useful in the log file

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/BaseClient.kt:
```diff
@@ -0,0 +1,28 @@
+
+    fun shutdown() = channel
+        .shutdown()
+        .awaitTermination(5, TimeUnit.SECONDS)
```
check and log staus of `awaitTermination` method call.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/BaseClient.kt:
```diff
@@ -0,0 +1,28 @@
+        .awaitTermination(5, TimeUnit.SECONDS)
+
+    protected fun waitForConnection() {
+        while (channel.getState(true) != ConnectivityState.READY) {
+            logger.info("Connecting to $hostname:$port...")
+            Thread.sleep(3_000)
+        }
+        logger.info("Connected.")
```
This check doesn't guarantee that a connection will be established when the next stub method is called.
Try to use CallOption().WithWaitForReady

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/MarketDataMonitor.kt:
```diff
@@ -0,0 +1,27 @@
+) : BaseClient(hostname, port, logger) {
+    fun run() {
+        waitForConnection()
+        stubRxAdapter.requestMarketDataStream(tickers).blockingSubscribe {
```
Add onError and onComplet handlers and log events details in them

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/MarketDataMonitor.kt:
```diff
@@ -0,0 +1,27 @@
+    fun run() {
+        waitForConnection()
+        stubRxAdapter.requestMarketDataStream(tickers).blockingSubscribe {
+            println(it.toString())
```
Use logging instead of a print method.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/MarketDataMonitor.kt:
```diff
@@ -0,0 +1,27 @@
+import com.github.lumber1000.simex.common.*
+
+fun main() {
+    val monitor = MarketDataMonitor(HOSTNAME, PORT, LOGGER, listOf("T1", "T2"))
```
Enter points have to catch an exception and log it otherwise we will have an investigation problem in the future.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -0,0 +1,60 @@
+
+fun main() {
+    val numberOfTraders = 4
+    val executor = Executors.newFixedThreadPool(numberOfTraders)
```
Could you please think about how to execute calls from different traders using only one thread?

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -0,0 +1,60 @@
+            val order = Order(0L, orderType, ticker, random.nextInt(100, 200), random.nextInt(1, 100), 0L)
+            stubRxAdapter.submitOrder(order)
+            logger.info("Send order: $order")
+            Thread.sleep(random.nextLong(200, 800))
```
Using sleep in executor's thread is bad practice, try to avoid this approach

Nikita-Smirnov-Exactpro replied:
The same advice is applicable to all blocking calls

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/StubRxAdapter.kt:
```diff
@@ -0,0 +1,80 @@
+import io.grpc.stub.StreamObserver
+import com.github.lumber1000.simex.common.*
+
+class StubRxAdapter(private val stub: SimexServiceGrpc.SimexServiceStub) {
```
Please reorder the definition of subclass and methods according to 
https://kotlinlang.org/docs/coding-conventions.html#class-layout

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/TestClient.kt:
```diff
@@ -0,0 +1,76 @@
+
+        // TODO: cancel streaming
+        stubRxAdapter.requestMarketDataStream(listOf("T1", "T2"))
+            .subscribe({ println("EVENT: $it") }, Throwable::printStackTrace)
```
Care about open-ended resources. Class or method which opens a resource should be responsible for its life cycle.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/TestClient.kt:
```diff
@@ -0,0 +1,76 @@
+                { id: Long ->
+                    order2.id = id
+                    logger.info("Order accepted: $order2")
+                }, Throwable::printStackTrace)
```
Please avoid code duplication.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Entities.kt:
```diff
@@ -0,0 +1,90 @@
+
+    val opposite get() = when(this) {
+        BUY_LIMIT -> SELL_LIMIT
+        SELL_LIMIT -> BUY_LIMIT
```
Pass opposite value to the constructor. In this case compiler guarantees that each new enum item will have the opposite value.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Entities.kt:
```diff
@@ -0,0 +1,90 @@
+
+import java.util.function.BiPredicate
+
+enum class OrderType(val acceptablePricePredicate: BiPredicate<Int, Int>) {
```
You can pass a lambda expression as argument to increase readability
for example val acceptablePricePredicate: (firstPrice: Int, secondPrice: Int) -> Boolean

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Entities.kt:
```diff
@@ -0,0 +1,90 @@
+    val timestamp: Long
+) {
+    fun isAcceptablePrice(otherOrder: Order) = type.acceptablePricePredicate.test(price, otherOrder.price)
+    override fun equals(other: Any?) = throw UnsupportedOperationException()
+    override fun hashCode() = throw UnsupportedOperationException()
```
Why do you use this code?

Nikita-Smirnov-Exactpro replied:
Maybe use id or ticker as a unique identifier?

Nikita-Smirnov-Exactpro replied:
Discussed this code and decided to make the price level entity with a queue to hold orders instead of the current implementation via TreeSet.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Entities.kt:
```diff
@@ -0,0 +1,90 @@
+        .setTimestamp(timestamp)
+
+    when (this) {
+        is Trade -> eventBuilder.tradeBuilder.setPrice(price).setSize(size)
+        is NewOrderAdded -> eventBuilder.orderAddedEventBuilder.setOrder(order.toMessage())
+        is OrderSizeChanged -> eventBuilder.orderSizeChangedBuilder.setNewSize(newSize)
```
Try to solve waring from the IDEA static analyzer or another

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Entities.kt:
```diff
@@ -0,0 +1,90 @@
+        is NewOrderAdded -> eventBuilder.orderAddedEventBuilder.setOrder(order.toMessage())
+        is OrderSizeChanged -> eventBuilder.orderSizeChangedBuilder.setNewSize(newSize)
+        is OrderRemoved -> { /* none */ }
+        else -> throw IllegalStateException("unknown event")
```
Use Kotlin's error method instead

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -0,0 +1,72 @@
+    private val orderBooks = HashMap<String, Book>()
+
+    private class Book {
+        private val orderSets = arrayOf(
```
EnumMap is more applicable to this case

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -0,0 +1,72 @@
+        operator fun get(type: OrderType): NavigableSet<Order> = orderSets[type.ordinal]
+    }
+
+    fun matchOrder(order: Order) {
```
This class should be thread-safe to handle parallel gRPC requests

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -0,0 +1,72 @@
+
+        if (order.size > 0) {
+            val isAdded = ordersBook[order.type].add(order)
+            check(isAdded)
```
Print error details and the context of execution.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -0,0 +1,72 @@
+import io.reactivex.rxjava3.subjects.PublishSubject
+import com.github.lumber1000.simex.common.*
+
+class OrderMatcher {
```
Users should be able to delete orders by id.
Matcher should be thread-safe to handle parallel requests

lumber1000 committed:
- Order book data structure includes order queues for every price level.
- Order book data structure allows removing orders by id (time complexity: O(1)).
- OrderMatcher class is thread safe.
- Traders execution implemented using Kotlin coroutines on a single thread executor. Blocking calls removed.
- Parameters (hostname and port) passed via config file (config.properties).
- CI via GitHub Actions.
- Kotlin version 1.5.32
- slf4j and log4j instead of java.util.Logger.
- Idea gradle plugin (adds generated files to source set).
- Gradle wrapper included.
- Default exception handler added.
- Code Duplication removed in TestClient.kt.

Nikita-Smirnov-Exactpro commented on build.gradle:
```diff
@@ -66,33 +66,37 @@ tasks.test {
     implementation "io.grpc:grpc-kotlin-stub:1.2.0"
 
+    implementation 'org.slf4j:slf4j-api:1.7.32'
+    implementation 'org.slf4j:slf4j-log4j12:1.7.32'
```
Please use the last version of log4j2

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -1,60 +1,79 @@
-    executor.shutdown()
-    executor.awaitTermination(1, TimeUnit.MINUTES)
-}
+import kotlinx.coroutines.*
+import java.util.concurrent.*
```
Configure your IDE to use explicit instead of multiple. This approach provides an increase speed class loading in huge projects

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -1,60 +1,79 @@
             stubRxAdapter.submitOrder(order)
-            logger.info("Send order: $order")
-            Thread.sleep(random.nextLong(200, 800))
+            LOGGER.info("Send order: $order")
```
Decrease log level to DEBUG for frequent operations. INFO log level is used for rare significant operations, for example, establishing a connection, changing state and etc

Nikita-Smirnov-Exactpro replied:
Use the lazy logging method to avoid redundantly String allocation when logging is off or has a higher log level than your call.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -1,60 +1,79 @@
-            var book = stubRxAdapter.getBook(listOf(ticker)).blockingGet()
-            println("\nOrderBook for $ticker (${book.size} orders)\n\n$book\n")
-            println("\n\nCancel all orders\n")
+        println("\nOrderBook for $ticker (${book.size} orders):\n\n$book\n")
+        println("Cancel unsatisfied orders:\n")
```
Use logger instead of print functions

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -1,60 +1,79 @@
 
-        barrier.await()
+    private suspend fun cancelUnsatisfiedOrders() {
+        var book = stubRxAdapter.getBook(listOf(ticker)).blockingGet()
```
This class is written via coroutine, try to use SimexServiceCoroutineStub instead of RxJava implemetation

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Config.kt:
```diff
@@ -1,14 +1,50 @@
 
-const val HOSTNAME = "localhost"
-const val PORT = 50051
+class Config (
+    val hostname: String,
+    val port: Int
+)
```
Use data class for bean classes

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/common/Config.kt:
```diff
@@ -1,14 +1,50 @@
+    val defaultHostname = "localhost"
+    val defaultPort = 50051
+
+    val properties = Properties()
```
Try to use the Jackson library to read config. It provides solve the same issue with a shorter code.

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/clients/RandomTrader.kt:
```diff
@@ -1,60 +1,79 @@
+                createTrader("T2", OrderType.SELL_LIMIT, 4000)
+            )
+
+            Executors.newSingleThreadExecutor().asCoroutineDispatcher().use { context ->
```
Redundant construction, because the runBlocking method already runs a new coroutine

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -1,72 +1,282 @@
+            while (!Thread.currentThread().isInterrupted) {
+                executeRequest(requestQueue.take())
+            }
+        } catch (_: InterruptedException) { /* exit */ }
```
Logging end execution event

Nikita-Smirnov-Exactpro commented on src/main/kotlin/com/github/lumber1000/simex/service/OrderMatcher.kt:
```diff
@@ -1,72 +1,282 @@
+    }
+
+    private fun executeRequest(request: Request) {
+        when (request.type) {
```
Add else block to prepare probably future enum extencions.
